consensus/leader.go

```diff
@@ -127,66 +122,89 @@ func (consensus *Consensus) onPrepare(msg *msg_pb.Message) {
 			Msg("[OnPrepare] No Matching Announce message")
 	}
 
-	prepareSig := recvMsg.Payload
-	prepareBitmap := consensus.prepareBitmap
+	//// Lock Read - Start
+	consensus.prepareMutex.RLock()
+
+	if !consensus.isRightBlockNumAndViewID(recvMsg) {
+		consensus.prepareMutex.RUnlock()
+		return
+	}
 
-	consensus.mutex.Lock()
-	defer consensus.mutex.Unlock()
+	blockHash := consensus.blockHash[:]
+	prepareBitmap := consensus.prepareBitmap
 	// proceed only when the message is not received before
 	signed := consensus.Decider.ReadBallot(quorum.Prepare, recvMsg.SenderPubkey.Bytes)
 	if signed != nil {
+		consensus.prepareMutex.RUnlock()
 		consensus.getLogger().Debug().
 			Str("validatorPubKey", recvMsg.SenderPubkey.Bytes.Hex()).
 			Msg("[OnPrepare] Already Received prepare message from the validator")
 		return
 	}
 
 	if consensus.Decider.IsQuorumAchieved(quorum.Prepare) {
+		consensus.prepareMutex.RUnlock()
 		// already have enough signatures
 		consensus.getLogger().Debug().
 			Str("validatorPubKey", recvMsg.SenderPubkey.Bytes.Hex()).
 			Msg("[OnPrepare] Received Additional Prepare Message")
 		return
 	}
+	signerCount := consensus.Decider.SignersCount(quorum.Prepare)
+	consensus.prepareMutex.RUnlock()
+	//// Lock Read - End
 
 	// Check BLS signature for the multi-sig
+	prepareSig := recvMsg.Payload
 	var sign bls.Sign
 	err = sign.Deserialize(prepareSig)
 	if err != nil {
 		consensus.getLogger().Error().Err(err).
 			Msg("[OnPrepare] Failed to deserialize bls signature")
 		return
 	}
-	if !sign.VerifyHash(recvMsg.SenderPubkey.Object, consensus.blockHash[:]) {
+	if !sign.VerifyHash(recvMsg.SenderPubkey.Object, blockHash) {
 		consensus.getLogger().Error().Msg("[OnPrepare] Received invalid BLS signature")
 		return
 	}
 
 	consensus.getLogger().Debug().
-		Int64("NumReceivedSoFar", consensus.Decider.SignersCount(quorum.Prepare)).
+		Int64("NumReceivedSoFar", signerCount).
 		Int64("PublicKeys", consensus.Decider.ParticipantsCount()).
 		Msg("[OnPrepare] Received New Prepare Signature")
+
+	//// Lock Write - Start
+	consensus.prepareMutex.Lock()
```

Comment:
As far as I know, this operation will introduce race condition which is not discoverable by --cover flag. Please compare the following two functions in multithreading execution:

```
type myStruct struct {
    val int
    mu sync.RWMutex
}

func (ms *myStruct) CompareAndSwap1(compareVal, targetVal int) {
    ms.mu.RLock()
    prev := ms.val
    ms.mu.RUnlock()

    if prev != compareVal {
        return
    }

    ms.mu.Lock()
    ms.val = targetVal  // At this point, ms.val can be already been changed by other goroutine
    ms.mu.Unlock()
}

// Use double lock check
func (ms *myStruct) CompareAndSwap1(compareVal, targetVal int) {
    ms.mu.RLock()
    prev := ms.val
    ms.mu.RUnlock()

    if prev == compareVal {
        ms.mu.Lock()
        if ms.val == compareVal {
            ms.val = targetVal    // value is protected. So this is fine
        }
        ms.mu.Unlock()
    }
}
```

If we really want to optimize in this code, I would recommend using the double lock checking.

Reply:
Good catch. Added the check of consistent block hash to make sure the read and write are operating on the same block. Note that checking on the same block is enough because the signature addition are always valid if it's the same block.